sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java

```diff
@@ -558,7 +574,23 @@ long flush(
               appendFailures.inc();
               return RetryType.RETRY_ALL_OPERATIONS;
             },
-            c -> recordsAppended.inc(c.protoRows.getSerializedRowsCount()),
+            c -> {
+              recordsAppended.inc(c.protoRows.getSerializedRowsCount());
+              if (successfulRowsReceiver != null) {
+                for (ByteString rowBytes : c.protoRows.getSerializedRowsList()) {
+                  try {
+                    TableRow row =
```

Comment:
This will restore the whole row and could be costly to just carry the whole thing around for the sole purpose of Wait.on(). Is it easy to return a row of a list of fields that can potentially be needed (something that constitutes a primary key of the record)?

Reply:
Some people have usages for this beyond just Wait.on (e.g. write row somewhere else only after first write is done), and the simples semantic is to return the entire row.

Agree about the cost - this is why I force users to opt into this. I think we should also output a collection of metadata (e.g. every append to the API along with some statistics, like the number of rows in that append), and this metadata collection can also be used in Wait.on.

Reply:
As long as docs will be clear about the windowing differences and the fact that the TableRow is not the same as the original (possibly different data types and missing fields that were ignored) this should work. 
Would be good to come up with the metadata approach that covers most use cases. Do you want to do it as part of this PR or as a separate case (with a separate BigQueryIO.Write method to indicate that this metadata collection should be produced?).

Reply:
Adding extra comments. IMO a metadata PCollection would be best done in a separate PR.